More secure path mangling

marcjansen opened the pull request:
This is a follow up from the excellent #280. I think the security might even be improved by the following PR.

I'd really like a review from @anikaweinmann or @mmacata on this; this is open for discussion.

marcjansen commented:
Hmm. Probably some occasions missed. Hopefully just that.

anikaweinmann commented on src/actinia_core/core/utils.py:
```diff
@@ -49,11 +50,22 @@ def os_path_normpath(path_parts):
+    # single location
+    allowed_starts = [
+        global_config.GRASS_USER_DATABASE,
+        global_config.GRASS_DATABASE,
```
Carmen is right that `GRASS_DATABASE` is only okay if read and not if you want to change something in it. Maybe you should still distinguish that somehow?

marcjansen replied:
Well it is used in [these lines of PR 280](https://github.com/mundialis/actinia_core/pull/280/files#diff-ee1d2bbb5ac0ad6fc98c8c9b1f98cb33232ee8d70320349b17362de44c629edcR217), that is why I included it.

anikaweinmann replied:
I am not sure if it makes sense to add some `r`/`w` parameter to the functions and split the `allowed_starts`.

marcjansen replied:
Ah, I now understand what you mean. I can update this as suggested.

marcjansen replied:
I have updated the PR; `ensure_valid_path` now accepts an optional second parameter, the intent ("r", "w" or "rw", Default is "r")

anikaweinmann commented on src/actinia_core/core/utils.py:
```diff
@@ -49,11 +50,22 @@ def os_path_normpath(path_parts):
+        global_config.GRASS_DATABASE,
+        global_config.GRASS_RESOURCE_DIR
+    ]
+    return any(path.startswith(allowed) for allowed in allowed_starts)
```
Theoretically the `startswith` could be unsafe here again, if  `path="GRASS_RESOURCE_DIR/../../something_else`.

anikaweinmann replied:
`os.path.normpath("GRASS_RESOURCE_DIR/../../something_else")` leads to `../something_else`, so the normpath is needed that this function `path_is_in_allowed_paths` is working. Maybe you can mention that path should be a path which was corrected with `os.path.normpath`.

marcjansen replied:
But if `os.path.normpath("GRASS_RESOURCE_DIR/../../something_else")` leads to `../something_else` the function would return false, as it does not start with any of the allowed prefixes. I can of course add a note

anikaweinmann commented:
@marcjansen do you want to add a `r`/`w` parameter to the functions? Otherwise it is also fine for me to merge it as it is.

marcjansen commented:
Sorry for not giving this more attention. I think the general idea is still valid, but I am not sure whether this catches all relevant cases. 

I see this has an approving review, and I can try to rebase this, if you still want it. I can also live with you guys and girls closing this PR with unmerged commits. Please go ahead as you see fit.

anikaweinmann commented:
@marcjansen you are welcome to merge and rebase

marcjansen commented:
> @marcjansen you are welcome to merge and rebase

I have done that and also added support for intents (`r`, `w` or `rw`) when `ensure_valid_path` is called.

Please have another close look at what is now in this PR. Where the method is called, I have added an intent where I was more or less sure, but these shpould please be double checked by someone with more know-how of the actual intent.

marcjansen commented:
anybody @mundialis-dev ?

marcjansen commented:
Friendly ping @mmacata @anikaweinmann

anikaweinmann commented:
Thank you, for me it look good. Please merge.

marcjansen commented:
Thanks for the review.
